File: csharp/Archive/Bricker_2.1/Game/Brick.cs
using Bricker.Rendering;
using SkiaSharp;
using System;

namespace Bricker.Game
{
    /// <summary>
    /// Represents a live, moving brick that has not yet joined the static game matrix.
    /// It will do so once it's hit bottom and come to rest.
    /// </summary>
    public class Brick
    {
        //private
        private readonly int _shapeNum;
        private readonly int _width;
        private readonly int _height;
        private readonly SKColor _color;

        private bool[,] _grid;
        private int _x;
        private int _y;
        private int _topSpace;
        private int _bottomSpace;
        private DateTime _lastDropTime;

        //public
        public int ShapeNum => _shapeNum;
        public int Width => _width;
        public int Height => _height;
        public SKColor Color => _color;
        public bool[,] Grid => _grid;
        public int X => _x;
        public int Y => _y;
        public int TopSpace => _topSpace;
        public int BottomSpace => _bottomSpace;
        public DateTime LastDropTime => _lastDropTime;

        /// <summary>
        /// Class constructor.
        /// </summary>
        public Brick(int shapeNum)
        {
            _shapeNum = shapeNum;
            switch (shapeNum)
            {
                case 1:
                    _width = 4;
                    _height = 4;
                    _grid = new bool[_width, _height];
                    _grid[0, 2] = true;
                    _grid[1, 2] = true;
                    _grid[2, 2] = true;
                    _grid[3, 2] = true;
                    _color = Colors.SilverPink;
                    break;
                case 2:
                    _width = 3;
                    _height = 3;
                    _grid = new bool[_width, _height];
                    _grid[0, 1] = true;
                    _grid[0, 2] = true;
                    _grid[1, 2] = true;
                    _grid[2, 2] = true;
                    _color = Colors.TuftsBlue;
                    break;
                case 3:
                    _width = 3;
                    _height = 3;
                    _grid = new bool[_width, _height];
                    _grid[2, 1] = true;
                    _grid[0, 2] = true;
                    _grid[1, 2] = true;
                    _grid[2, 2] = true;
                    _color = Colors.ChromeYellow;
                    break;
                case 4:
                    _width = 2;
                    _height = 2;
                    _grid = new bool[_width, _height];
                    _grid[0, 0] = true;
                    _grid[0, 1] = true;
                    _grid[1, 0] = true;
                    _grid[1, 1] = true;
                    _color = Colors.Independence;
                    break;
                case 5:
                    _width = 4;
                    _height = 4;
                    _grid = new bool[_width, _height];
                    _grid[1, 0] = true;
                    _grid[2, 0] = true;
                    _grid[0, 1] = true;
                    _grid[1, 1] = true;
                    _color = Colors.ForestGreen;
                    break;
                case 6:
                    _width = 4;
                    _height = 4;
                    _grid = new bool[_width, _height];
                    _grid[1, 1] = true;
                    _grid[0, 2] = true;
                    _grid[1, 2] = true;
                    _grid[2, 2] = true;
                    _color = Colors.Byzantine;
                    break;
                case 7:
                    _width = 4;
                    _height = 4;
                    _grid = new bool[_width, _height];
                    _grid[0, 0] = true;
                    _grid[1, 0] = true;
                    _grid[1, 1] = true;
                    _grid[2, 1] = true;
                    _color = Colors.Coquelicot;
                    break;
            }
            _topSpace = CalculateTopSpace();
            _bottomSpace = CalculateBottomSpace();
            _x = (12 - _width) / 2;
            _y = 1 - _topSpace;
            _lastDropTime = DateTime.Now;
        }

        /// <summary>
        /// Gets spacing at top of brick.
        /// </summary>
        private int CalculateTopSpace()
        {
            int topSpace = 0;
            for (int y = 0; y < _height; y++)
            {
                bool empty = true;
                for (int x = 0; x < _width; x++)
                {
                    if (_grid[x, y])
                        empty = false;
                }
                if (empty)
                    topSpace++;
                else
                    break;
            }
            return topSpace;
        }

        /// <summary>
        /// Gets spacing at bottom of brick.
        /// </summary>
        private int CalculateBottomSpace()
        {
            int bottomSpace = 0;
            for (int y = _height - 1; y >= 0; y--)
            {
                bool empty = true;
                for (int x = 0; x < _width; x++)
                {
                    if (_grid[x, y])
                        empty = false;
                }
                if (empty)
                    bottomSpace++;
                else
                    break;
            }
            return bottomSpace;
        }

        /// <summary>
        /// Returns true if filled brick space conflicts with filled matrix space.
        /// </summary>
        public bool Collision(bool[,] matrix)
        {
            for (int x = 0; x < _width; x++)
            {
                for (int y = 0; y < _height; y++)
                {
                    if (_grid[x, y])
                    {
                        int mX = x + _x;
                        int mY = y + _y;
                        if ((mX < 0) || (mX > 21))
                            return true;
                        if ((mY < 0) || (mY > 21))
                            return true;
                        if (matrix[mX, mY])
                            return true;
                    }
                }
            }
            return false;
        }

        /// <summary>
        /// Moves the brick left, unless collision.
        /// </summary>
        public void MoveLeft(bool[,] matrix)
        {
            _x--;
            if (Collision(matrix))
                _x++;
        }

        /// <summary>
        /// Moves the brick right, unless collision.
        /// </summary>
        public void MoveRight(bool[,] matrix)
        {
            _x++;
            if (Collision(matrix))
                _x--;
        }

        /// <summary>
        /// Moves the brick down, unless collision.
        /// </summary>
        public bool MoveDown(bool[,] matrix)
        {
            _lastDropTime = DateTime.Now;
            _y++;
            if (Collision(matrix))
            {
                _y--;
                return true;
            }
            return false;
        }

        /// <summary>
        /// Returns true if its time to drop brick (gravity).
        /// </summary>
        public bool IsDropTime(double intervalMs)
        {
            double elapsedMs = (DateTime.Now - _lastDropTime).TotalMilliseconds;
            bool dropTime = elapsedMs >= intervalMs;
            return dropTime;
        }

        /// <summary>
        /// Rotates the brick 90* clockwise, moving slightly if there's a collision.
        /// </summary>
        public void Rotate(bool[,] matrix)
        {
            bool[,] newGrid = new bool[_height, _width];
            for (int x1 = 0; x1 < _width; x1++)
            {
                for (int y1 = 0; y1 < _height; y1++)
                {
                    int x2 = -y1 + (_height - 1);
                    int y2 = x1;
                    newGrid[x2, y2] = _grid[x1, y1];
                }
            }
            _grid = newGrid;

            int steps = 0;
            while (Collision(matrix))
            {
                _y++;
                steps++;
                if (steps >= 3)
                {
                    _y -= 3;
                    break;
                }
            }

            steps = 0;
            while (Collision(matrix))
            {
                _y--;
                steps++;
                if (steps >= 3)
                {
                    _y += 3;
                    break;
                }
            }

            steps = 0;
            while (Collision(matrix))
            {
                _x++;
                steps++;
                if (steps >= 3)
                {
                    _x -= 3;
                    break;
                }
            }

            steps = 0;
            while (Collision(matrix))
            {
                _x--;
                steps++;
                if (steps >= 3)
                {
                    _x += 3;
                    break;
                }
            }

            _topSpace = CalculateTopSpace();
            _bottomSpace = CalculateBottomSpace();
        }

    }
}

File: python/bricker/brick.py
"""Bricker - A Tetris-like brick game.
Copyright (C) 2017-2020  John Hyland
GNU GENERAL PUBLIC LICENSE Version 3"""

from typing import List
from time import perf_counter
from color import Colors, Color


class Brick:
    """Represents a live, moving brick that has not yet joined the static game matrix.
    It will do so once it's hit bottom and come to rest."""

    def __init__(self, shape_num: int) -> None:
        """Class constructor.  Creates one of seven basic shapes."""
        self.__shape_num: int = shape_num
        if shape_num == 1:
            self.__width: int = 4
            self.__height: int = 4
            self.__grid: List[List[int]] = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[0][2] = 1
            self.__grid[1][2] = 1
            self.__grid[2][2] = 1
            self.__grid[3][2] = 1
            self.__color = Colors.SilverPink
        elif shape_num == 2:
            self.__width = 3
            self.__height = 3
            self.__grid = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[0][1] = 1
            self.__grid[0][2] = 1
            self.__grid[1][2] = 1
            self.__grid[2][2] = 1
            self.__color = Colors.TuftsBlue
        elif shape_num == 3:
            self.__width = 3
            self.__height = 3
            self.__grid = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[2][1] = 1
            self.__grid[0][2] = 1
            self.__grid[1][2] = 1
            self.__grid[2][2] = 1
            self.__color = Colors.ChromeYellow
        elif shape_num == 4:
            self.__width = 2
            self.__height = 2
            self.__grid = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[0][0] = 1
            self.__grid[0][1] = 1
            self.__grid[1][0] = 1
            self.__grid[1][1] = 1
            self.__color = Colors.Independence
        elif shape_num == 5:
            self.__width = 3
            self.__height = 3
            self.__grid = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[1][0] = 1
            self.__grid[2][0] = 1
            self.__grid[0][1] = 1
            self.__grid[1][1] = 1
            self.__color = Colors.ForestGreen
        elif shape_num == 6:
            self.__width = 3
            self.__height = 3
            self.__grid = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[1][1] = 1
            self.__grid[0][2] = 1
            self.__grid[1][2] = 1
            self.__grid[2][2] = 1
            self.__color = Colors.Byzantine
        elif shape_num == 7:
            self.__width = 3
            self.__height = 3
            self.__grid = [[0 for x in range(self.__height)] for y in range(self.__width)]
            self.__grid[0][0] = 1
            self.__grid[1][0] = 1
            self.__grid[1][1] = 1
            self.__grid[2][1] = 1
            self.__color = Colors.Coquelicot
        self.__top_space: int = self.__get_top_space()
        self.__bottom_space: int = self.__get_bottom_space()
        self.__x: int = int((12 - self.__width) / 2)
        self.__y: int = 1 - self.__top_space
        self.__last_drop_time: float = perf_counter()

    @property
    def shape_num(self) -> int:
        """Returns the shape number."""
        return self.__shape_num

    @property
    def width(self) -> int:
        """Returns brick width."""
        return self.__width

    @property
    def height(self) -> int:
        """Returns brick height."""
        return self.__height

    @property
    def grid(self) -> List[List[int]]:
        """Returns brick grid."""
        return self.__grid

    @property
    def color(self) -> Color:
        """Returns brick color."""
        return self.__color

    @property
    def top_space(self) -> int:
        """Returns non-solid spaces at top of brick grid."""
        return self.__top_space

    @property
    def bottom_space(self) -> int:
        """Returns non-solid spaces at bottom of brick grid."""
        return self.__bottom_space

    @property
    def x(self) -> int:
        """Returns X position of brick."""
        return self.__x

    @property
    def y(self) -> int:
        """Returns Y position of brick."""
        return self.__y

    def __get_top_space(self) -> int:
        """Calculates non-solid spaces at top of brick grid."""
        top_space = 0
        for y in range(0, self.__height):
            empty = True
            for x in range(0, self.__width):
                if self.__grid[x][y] == 1:
                    empty = False
            if empty:
                top_space += 1
            else:
                break
        return top_space

    def __get_bottom_space(self) -> int:
        """Calculates non-solid spaces at bottom of brick grid."""
        bottom_space = 0
        for y in reversed(range(0, self.__height)):
            empty = True
            for x in range(0, self.__width):
                if self.__grid[x][y] == 1:
                    empty = False
            if empty:
                bottom_space += 1
            else:
                break
        return bottom_space

    def collision(self, matrix) -> bool:
        """Returns true on brick collision."""
        for x in range(0, self.__width):
            for y in range(0, self.__height):
                matrix_x = x + self.__x
                matrix_y = y + self.__y
                if (self.__grid[x][y] == 1) and (matrix[matrix_x][matrix_y] == 1):
                    return True
        return False

    def move_left(self, matrix) -> None:
        """Moves brick left, prevents collision."""
        self.__x -= 1
        if self.collision(matrix):
            self.__x += 1

    def move_right(self, matrix) -> None:
        """Moves brick right, prevents collision."""
        self.__x += 1
        if self.collision(matrix):
            self.__x -= 1

    def move_down(self, matrix) -> bool:
        """Moves brick down, prevents collision.  Returns true if move would have hit bottom."""
        self.__last_drop_time = perf_counter()
        self.__y += 1
        if self.collision(matrix):
            self.__y -= 1
            return True
        return False

    def is_drop_time(self, interval) -> bool:
        """Returns true if its time to drop brick (gravity)."""
        now = perf_counter()
        elapsed = now - self.__last_drop_time
        drop_time = elapsed >= interval
        return drop_time

    def rotate(self, matrix) -> None:
        """Rotates brick."""

        new_grid = [[0 for x in range(self.__width)] for y in range(self.__height)]
        for x1 in range(0, self.__width):
            for y1 in range(0, self.__height):
                x2 = -y1 + (self.__height - 1)
                y2 = x1
                new_grid[x2][y2] = self.__grid[x1][y1]
        self.__grid = new_grid

        steps = 0
        while self.collision(matrix):
            self.__y += 1
            steps += 1
            if steps >= 3:
                self.__y -= 3
                break

        steps = 0
        while self.collision(matrix):
            self.__y -= 1
            steps += 1
            if steps >= 3:
                self.__y += 3
                break

        steps = 0
        while self.collision(matrix):
            self.__x -= 1
            steps += 1
            if steps >= 3:
                self.__x += 3
                break

        steps = 0
        while self.collision(matrix):
            self.__x += 1
            steps += 1
            if steps >= 3:
                self.__x -= 3
                break
